flash_io.c

```diff
@@ -0,0 +1,197 @@
+  buf[1] = targets;
+
+  memcpy((void*)&buf[2], (void*)flash_io, (size_t)FLASH_USER_DATA_LEN);
+
+  check_end_of_user_flash();
+
+  uint32_t ints = save_and_disable_interrupts();
+  flash_range_program((uint32_t)flash_ptr - XIP_BASE, buf, FLASH_PAGE_SIZE);
+  restore_interrupts(ints);
+
+  flash_ptr += FLASH_PAGE_SIZE;
+}
+
+static int
+flash_compare_diffs(furnace_context_t* ctx)
+{
+  ctx->flash_io.des_temp   = ctx->pilot.des_temp;
+  ctx->flash_io.is_enabled = ctx->pilot.is_enabled;
+  ctx->flash_io.log_bits   = ctx->log_bits;
+
+#if CONFIG_WATER
+  ctx->flash_io.pwm_water  = ctx->pwm_water;
+#endif
+
+  bool diff = memcmp(&ctx->flash_io, &ctx->flash_last_written, FLASH_USER_DATA_LEN);
+
+  if(diff != 0)
+    memcpy(&(ctx->flash_last_written), &(ctx->flash_io), FLASH_USER_DATA_LEN);
+
+ return diff;
+}
+
+static void
+read_flash_data(furnace_context_t* ctx)
+{
+  const void* tmp = flash_ptr;
+
+  // skip size (uint8_t) and targets
+  tmp += sizeof(uint8_t);
+  tmp += sizeof(enum init_targets);
+
+  memcpy(&(ctx->flash_io), tmp, FLASH_USER_DATA_LEN);
+  memcpy(&(ctx->flash_last_written), &(ctx->flash_io), FLASH_USER_DATA_LEN);
+
+  ctx->pilot.des_temp      = ctx->flash_io.des_temp;
+  ctx->pilot.is_enabled    = ctx->flash_io.is_enabled;
+  ctx->log_bits            = ctx->flash_io.log_bits;
+
+#if CONFIG_WATER
+  ctx->pwm_water = ctx->flash_io.pwm_water;
+#endif
+}
+
+static void
+check_for_target(const void** last_correct_page)
+{
+  const void* tmp = flash_ptr;
+  tmp += sizeof(uint8_t);
+  const enum init_targets saved_targets = *(enum init_targets*)tmp;
+
+  if(targets == saved_targets)
+    *last_correct_page = flash_ptr;
+  else
+    *last_correct_page = NULL;
+}
+
+static void
+init_flash(furnace_context_t* ctx)
+{
+  // There is little chance for us to ever reach 256 (FLASH_PAGE_SIZE) bytes of
+  // size of flash_io_t.
+  // Also, 255 is reserved.
+  // We need to distinct the difference between size and empty memory block.
+  //
+  // Although, if we pass this number, we must get compile-time error.
+  // In such a case, this algorithm is useless,
+  // and we will need to develop a new one.
+  //
+  // We assume that sizeof(enum init_targets) is 1 byte.
+  static_assert(FLASH_USER_DATA_LEN < FLASH_PAGE_SIZE - 2, "flash_io_t is too large\n");
+  static_assert(sizeof(enum init_targets) == 1, "init_targets size is different than assumed\n");
+
+  flash_ptr = &FLASH_USER_DATA_START;
+  uint8_t size;
+  const void* last_correct_page = NULL;
+
+  while(true)
+  {
+    if(check_end_of_user_flash())
+      break;
+
+    size = *(uint8_t*)flash_ptr;
+
+    if(size == FLASH_USER_DATA_LEN)
+      check_for_target(&last_correct_page);
+    else if((int8_t)size == -1)
+    {
+      if(last_correct_page != NULL)
+      {
+        flash_ptr = last_correct_page;
+        read_flash_data(ctx);
+        flash_ptr += FLASH_PAGE_SIZE;
+      }
+
+      break;
+    }
+    else
+      last_correct_page = NULL;
+
+    flash_ptr += FLASH_PAGE_SIZE;
+  }
```

Comment:
I don't really like this whole algorithm. I think It's just too complex for what we actually want to do.

We just need a always advancing counter that can hold `id > NUM_FLASH_SLOTS`, where `NUM_FLASH_SLOTS` is a number of slots where the saved data structure can be saved into.

Consider this:
We have 4k flash buffer for saving the non-volatile data.
`FLASH_PAGE_SIZE` is 256 bytes, so we get `NUM_FLASH_SLOTS = 4096 / FLASH_PAGE_SIZE`.
Every time we save a structure, we increment the `id` inside it.
The most recent structure (index R) is the one for which this **doesn't** hold:
`flash_slots[R].id == flash_slots[(R + 1) % NUM_FLASH_SLOTS].id + 1`